### PR TITLE
Attempt building for macOS 10.14 Mojave

### PR DESCRIPTION
**Pull request by jasongrout**
This is a bit of a shot in the dark, but it may be that we just need to set an environment variable to configure building for an older version of macOS. See https://github.com/johnboiles/obs-mac-virtualcam/issues/85#issuecomment-628298465

### CODE REVIEW

**Comment by johnboiles** [comment]
Can't hurt to try the simple thing first!

But it [looks like](https://github.com/johnboiles/obs-mac-virtualcam/runs/672698220#step:5:127) the build is using the 10.15 SDK.

I think that [SO post](https://stackoverflow.com/questions/58642779/compile-with-gcc-on-macos-10-14-5-that-will-support-earlier-versions-of-mac/58653850#58653850) you found is very promising. Probably we will need a build step to download and unzip the older SDK from [here](https://github.com/phracker/MacOSX-SDKs/releases) (let's start with 10.14 before trying 10.13).

If you are able to keep jamming on this, maybe make a new step in the [build.yml](https://github.com/johnboiles/obs-mac-virtualcam/blob/master/.github/workflows/build.yml) that downloads and unzips the 10.14 SDK then does a `ls /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs` to confirm that it's in place.

**Comment by johnboiles** [comment]
I also saw this. Not sure if we'll need to do it

image

**Comment by jasongrout** [comment]
Yes, I was thinking the same thing. Interestingly, I only seem to have the 10.15 sdk on my 10.14 computer, though (apparently xcode 11.3 only comes with the 10.15 sdk?). Reading [this](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/cross_development/Configuring/configuring.html#//apple_ref/doc/uid/10000163i-CH1-SW1) made me hopeful that perhaps it might work to set the minimum version and use the 10.15 sdk.

I couldn't find an official place to download the 10.14 sdk - that's why I thought I'd try just the environment variable first.

**Comment by johnboiles** [comment]
Oh very interesting that you can build on your 10.14 machine with only the 10.15 SDK! I totally get your line of reasoning in that case. Maybe these artifacts you're generating will just work 🤞

**Comment by jasongrout** [comment]
> If you are able to keep jamming on this, maybe make a new step in the [build.yml](https://github.com/johnboiles/obs-mac-virtualcam/blob/master/.github/workflows/build.yml) that downloads and unzips the 10.14 SDK then does a `ls /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs` to confirm that it's in place. Maybe also do a `ls` before putting the SDK in place; who knows maybe GitHub's action VM already has this in place.

Plus add that sdk to a cache so that we don't have to download it each time a PR comes in :).

**Comment by johnboiles** [comment]
> Plus add that sdk to a cache

😄 yes please!

**Comment by jasongrout** [comment]
> Oh very interesting that you can build on your 10.14 machine with only the 10.15 SDK!

(Really, I don't know what I'm doing here, so take everything I say with a huge boulder of salt)

**Comment by johnboiles** [comment]
So if you `ls /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs` on your machine you see only the `10.15` SDK?

I'm def guessing here too. I've dealt with version compatibility on iOS in the past, but that was just clicking some things in Xcode so I have no idea what it was doing under the hood.

**Comment by jasongrout** [comment]
How exactly do I install the plugin artifacts into my stock install of OBS? I see the `.plugin` directory goes in `/Library/CoreMediaIO/Plug-Ins/DAL`. Where does the `.so` go?

**Comment by jasongrout** [comment]
```
% ls -las /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs
total 0
0 drwxr-xr-x  5 root  wheel  160 Jan 13 20:33 ./
0 drwxr-xr-x  6 root  wheel  192 Jan 13 20:33 ../
0 drwxr-xr-x  4 root  wheel  128 Jan 13 20:32 DriverKit19.0.sdk/
0 drwxr-xr-x  8 root  wheel  256 Jan 13 20:33 MacOSX.sdk/
0 lrwxr-xr-x  1 root  wheel   10 Sep 26  2019 MacOSX10.15.sdk@ -> MacOSX.sdk
```

**Comment by johnboiles** [comment]
By default it's in `/Library/Application Support/obs-studio/plugins`. So you can

    sudo cp obs-mac-virtualcam.so /Library/Application\ Support/obs-studio/plugins/obs-mac-virtualcam/bin/obs-mac-virtualcam.so

If that has problems running because it isn't signed, we've [found](https://github.com/johnboiles/obs-mac-virtualcam/issues/23#issuecomment-626890806) that we can just dump the .so into application bundle plugins dir and it seems to just work without needing codesigning. `sudo cp obs-mac-virtualcam.so /Applications/OBS.app/Contents/PlugIns/`

**Comment by jasongrout** [comment]
Thanks. It works!!! I used the plugin artifacts generated in this build workflow, copied them to the locations you said, and it works as well as it did when I compiled myself.

**Comment by jasongrout** [comment]
(which for me means that Chrome was able to see the virtual camera and the content of it, and quicktime was able to see the virtual camera, but only displayed the test card, not the actual content)

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ jobs:
         cd obs-studio/build
 
         export QTDIR=$(brew --prefix qt5)
+        export MACOSX_DEPLOYMENT_TARGET=10.14
```
Comment (johnboiles) [inline .github/workflows/build.yml:34]: I _think_ we can set this as an environment variable for the entire job like this.

```
jobs:
  build:
    env:
      MACOSX_DEPLOYMENT_TARGET: 10.14
    runs-on: macos-latest
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -39,8 +40,32 @@ jobs:
...
         cd obs-mac-virtualcam/build
 
+        export MACOSX_DEPLOYMENT_TARGET=10.14
         cmake -DLIBOBS_INCLUDE_DIR:STRING=$GITHUB_WORKSPACE/obs-studio/libobs \
           -DLIBOBS_LIB:STRING=$GITHUB_WORKSPACE/obs-studio/build/libobs/libobs.dylib \
           -DOBS_FRONTEND_LIB:STRING=$GITHUB_WORKSPACE/obs-studio/build/UI/obs-frontend-api/libobs-frontend-api.dylib \
           -DQTDIR:STRING=/usr/local/opt/qt ..
         make -j
+
+    - name: Fix runtime QT deps
```
Comment (johnboiles) [inline .github/workflows/build.yml:50]: I'm fine leaving this. Probably useful to be able to test PRs

**Comment by johnboiles** [comment]
Hey that's awesome! So this fully works with Chrome? Can you see if you can find a Console.app log for QuickTime. Maybe it's a codesigning thing for Quicktime and it will just work when we sign this with a production cert.

File: .github/workflows/build.yml (hunk 1 above)
Comment (jasongrout) [reply .github/workflows/build.yml]: Right. Done. Thanks.

**Comment by jasongrout** [comment]
Here's the relevant part from the quicktime log when I switch to the virtual camera:

```
2020-05-13 17:40:38.286 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:38.287 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:38.287 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:38.287 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.548 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.548 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.549 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyHogMode self=0x10eba4590 settable=0
2020-05-13 17:40:39.549 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyExcludeNonDALAccess self=0x10eba4590 settable=1
2020-05-13 17:40:39.549 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyExcludeNonDALAccess self=0x10eba4590 settable=1
2020-05-13 17:40:39.549 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIODevicePropertyExcludeNonDALAccess self=0x10eba4590 data(int)=1
2020-05-13 17:40:39.549 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.551 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.551 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.551 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.552 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLocation
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 data(int)=66110
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: Stream unhandled isPropertySettableWithAddress for kCMIOStreamPropertyFormatDescription
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Stream(45) kCMIOStreamPropertyFormatDescription self=0x10eba4590 settable=0
2020-05-13 17:40:39.762 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 data(int)=-1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 data(int)=66110
2020-05-13 17:40:39.763 QuickTime Player[66110:541977] OBSDAL: Stream unhandled isPropertySettableWithAddress for kCMIOStreamPropertyFrameRate
2020-05-13 17:40:39.764 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Stream(45) kCMIOStreamPropertyFrameRate self=0x10eba4590 settable=0
2020-05-13 17:40:39.764 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.764 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.764 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectIsPropertySettable Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 settable=1
2020-05-13 17:40:39.764 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIODevicePropertyDeviceMaster self=0x10eba4590 data(int)=-1
2020-05-13 17:40:39.835 QuickTime Player[66110:541977] OBSDAL: Device unhandled hasPropertyWithAddress for kCMIODevicePropertyLinkedAndSyncedCoreAudioDeviceUID
2020-05-13 17:40:39.836 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_ObjectSetPropertyData Device(44) kCMIOObjectPropertyListenerAdded self=0x10eba4590 data(int)=1818850926
2020-05-13 17:40:39.836 QuickTime Player[66110:541977] OBSDAL: Device unhandled setPropertyDataWithAddress for kCMIOObjectPropertyListenerAdded
2020-05-13 17:40:39.846 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_StreamCopyBufferQueue <Stream: 0x7fe6abc90c80> (id=45) self=0x10eba4590 queue=<CMSimpleQueue 0x7fe6ad8a08d0 retainCount: 2 allocator: 0x7fffaeeec8e0
	capacity = 30
	count = 1
	readIndex = 28
	writeIndex = 29
	elements[readIndex..(writeIndex-1) %% (capacity+1)] = {
		0x7fe6ad92c840}
>
2020-05-13 17:40:39.846 QuickTime Player[66110:541977] OBSDAL: HardwarePlugIn_DeviceStartStream self=0x10eba4590 device=44 stream=45
2020-05-13 17:40:39.846 QuickTime Player[66110:541977] OBSDAL: -[PlugIn startStream] 
2020-05-13 17:40:39.847 QuickTime Player[66110:542181] OBSDAL: -[Stream startServingDefaultFrames] 
2020-05-13 17:40:39.847 QuickTime Player[66110:542181] OBSDAL: Server is not available
GVA encoder info: recomputed for fps: 30.000000, gop size in pics: 30, gop size in sec: 1.000000 
2020-05-13 17:40:40.134 QuickTime Player[66110:542048] OBSDAL: Server is not available
2020-05-13 17:40:41.134 QuickTime Player[66110:542683] OBSDAL: Server is not available
2020-05-13 17:40:42.133 QuickTime Player[66110:542181] OBSDAL: Server is not available
2020-05-13 17:40:43.133 QuickTime Player[66110:542048] OBSDAL: Server is not available
2020-05-13 17:40:44.133 QuickTime Player[66110:542179] OBSDAL: Server is not available
2020-05-13 17:40:45.134 QuickTime Player[66110:542684] OBSDAL: Server is not available
2020-05-13 17:40:46.134 QuickTime Player[66110:542682] OBSDAL: Server is not available
2020-05-13 17:40:47.133 QuickTime Player[66110:542048] OBSDAL: Server is not available
```
Then it just keeps repeating that last line about once a second.

**Comment by jasongrout** [comment]
> So this fully works with Chrome?

Yes, it appears to work just fine with Chrome for me.

**Comment by johnboiles** [comment]
Very interesting. I wonder if there's some differences in QuickTime's ability to reach OBS over mach ipc in 10.14

**Comment by johnboiles** [comment]
This will have the pkg when its done https://github.com/johnboiles/obs-mac-virtualcam/actions/runs/104162509

**Comment by Mesihas** [comment]
Works !!! on my mojave. Cheers!!

**Comment by jasongrout** [comment]
That pkg also works for me. Thanks again!

**Comment by johnboiles** [comment]
Thata's awesome!! I wonder if we can just dial this down to 10.13 for the high sierra people.

I'll cut a 1.1.0 release from this when i get a chance!!

**Comment by denwald** [comment]
> This will have the pkg when its done https://github.com/johnboiles/obs-mac-virtualcam/actions/runs/104162509

This works for me on Mojave with Quicktime as the video-sink as well! First it only showed the test screen. But after a reboot I can see live video in Quicktime. Maybe that's because I had the Catalina pkg already installed and then used the Mojave installer to replace the plugin.

**Pull request merged**